File: Mars-Images/CameraModel.m
//
//  CameraModel.m
//  Mars-Images
//
//  Created by Mark Powell on 2/4/14.
//  Copyright (c) 2014 Powellware. All rights reserved.
//

#import "CameraModel.h"
#import "CAHV.h"
#import "CAHVOR.h"
#import "CAHVORE.h"

@implementation CameraModel

- (NSArray*) size {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:[NSString stringWithFormat:@"You must override %@ in a subclass", NSStringFromSelector(_cmd)]
                                 userInfo:nil];
}

+ (id<Model>) model: (NSArray*) modelJSON {
    id<Model> returnedModel = nil;
    NSString* type;
    NSArray *c, *a, *h, *v, *o, *r, *e;
    NSNumber *mtype, *mparm;
    NSNumber *width, *height;
    for (NSObject* obj in modelJSON) {
        if ([obj isKindOfClass:[NSDictionary class]]) {
            NSDictionary* dict = (NSDictionary*) obj;
            for (NSString* key in dict.allKeys) {
                if ([key isEqualToString:@"type"]) {
                    type = [dict objectForKey:key];
                } else if ([key isEqualToString:@"components"]) {
                    NSDictionary* comps = [dict objectForKey:key];
                    c = [comps objectForKey:@"c"];
                    a = [comps objectForKey:@"a"];
                    h = [comps objectForKey:@"h"];
                    v = [comps objectForKey:@"v"];
                    o = [comps objectForKey:@"o"];
                    r = [comps objectForKey:@"r"];
                    e = [comps objectForKey:@"e"];
                    mtype = [comps objectForKey:@"t"];
                    mparm = [comps objectForKey:@"p"];
                } else if ([key isEqualToString:@"area"]) {
                    NSArray* area = [dict objectForKey:key];
                    width = [area objectAtIndex:0];
                    height = [area objectAtIndex:1];
                }
            }
        }
    }
    
    if (!type) {
        NSLog(@"Brown alert: camera model type not found.");
        return NULL;
    } else if ([type isEqualToString:@"CAHV"]) {
        CAHV* model = [[CAHV alloc] init];
        [model setC: ((NSNumber*)[c objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[c objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[c objectAtIndex:2]).floatValue];
        [model setA: ((NSNumber*)[a objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[a objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[a objectAtIndex:2]).floatValue];
        [model setH: ((NSNumber*)[h objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[h objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[h objectAtIndex:2]).floatValue];
        [model setV: ((NSNumber*)[v objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[v objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[v objectAtIndex:2]).floatValue];
        returnedModel = model;
    } else if ([type isEqualToString:@"CAHVOR"]) {
        CAHVOR* model = [[CAHVOR alloc] init];
        [model setC: ((NSNumber*)[c objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[c objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[c objectAtIndex:2]).floatValue];
        [model setA: ((NSNumber*)[a objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[a objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[a objectAtIndex:2]).floatValue];
        [model setH: ((NSNumber*)[h objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[h objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[h objectAtIndex:2]).floatValue];
        [model setV: ((NSNumber*)[v objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[v objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[v objectAtIndex:2]).floatValue];
        [model setO: ((NSNumber*)[o objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[o objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[o objectAtIndex:2]).floatValue];
        [model setR: ((NSNumber*)[r objectAtIndex:0]).floatValue
                  y: ((NSNumber*)[r objectAtIndex:1]).floatValue
                  z: ((NSNumber*)[r objectAtIndex:2]).floatValue];
        returnedModel = model;
   } else if ([type isEqualToString:@"CAHVORE"]) {
       CAHVORE* model = [[CAHVORE alloc] init];
       [model setC: ((NSNumber*)[c objectAtIndex:0]).floatValue
                 y: ((NSNumber*)[c objectAtIndex:1]).floatValue
                 z: ((NSNumber*)[c objectAtIndex:2]).floatValue];
       [model setA: ((NSNumber*)[a objectAtIndex:0]).floatValue
                 y: ((NSNumber*)[a objectAtIndex:1]).floatValue
                 z: ((NSNumber*)[a objectAtIndex:2]).floatValue];
       [model setH: ((NSNumber*)[h objectAtIndex:0]).floatValue
                 y: ((NSNumber*)[h objectAtIndex:1]).floatValue
                 z: ((NSNumber*)[h objectAtIndex:2]).floatValue];
       [model setV: ((NSNumber*)[v objectAtIndex:0]).floatValue
                 y: ((NSNumber*)[v objectAtIndex:1]).floatValue
                 z: ((NSNumber*)[v objectAtIndex:2]).floatValue];
       [model setO: ((NSNumber*)[o objectAtIndex:0]).floatValue
                 y: ((NSNumber*)[o objectAtIndex:1]).floatValue
                 z: ((NSNumber*)[o objectAtIndex:2]).floatValue];
       [model setR: ((NSNumber*)[r objectAtIndex:0]).floatValue
                 y: ((NSNumber*)[r objectAtIndex:1]).floatValue
                 z: ((NSNumber*)[r objectAtIndex:2]).floatValue];
       [model setE: ((NSNumber*)[e objectAtIndex:0]).floatValue
                 y: ((NSNumber*)[e objectAtIndex:1]).floatValue
                 z: ((NSNumber*)[e objectAtIndex:2]).floatValue];
       [model setT: mtype.floatValue];
       [model setP: mparm.floatValue];
       returnedModel = model;
    }
    if (returnedModel) {
        [((CAHV*)returnedModel) setXdim: width.intValue];
        [((CAHV*)returnedModel) setYdim: height.intValue];
    }
    return returnedModel;
}

+ (NSArray*) origin: (NSArray*) modelJson {
    for (NSObject* obj in modelJson) {
        if ([obj isKindOfClass:[NSDictionary class]]) {
            NSDictionary* dict = (NSDictionary*) obj;
            for (NSString* key in dict.allKeys) {
                if ([key isEqualToString:@"origin"]) {
                    NSArray* comps = [dict objectForKey:key];
                    NSNumber* x = [comps objectAtIndex:0];
                    NSNumber* y = [comps objectAtIndex:1];
                    return [NSArray arrayWithObjects: x, y, nil];
                }
            }
        }
    }
    NSLog(@"Brown alert: origin not found in camera model.");
    return [[NSArray alloc] init];
}

+ (NSArray*) pointingVector: (NSArray*) modelJson {
    for (NSObject* obj in modelJson) {
        if ([obj isKindOfClass:[NSDictionary class]]) {
            NSDictionary* dict = (NSDictionary*) obj;
            for (NSString* key in dict.allKeys) {
                if ([key isEqualToString:@"camera_vector"]) {
                    NSArray* comps = [dict objectForKey:key];
                    NSNumber* x = [comps objectAtIndex:0];
                    NSNumber* y = [comps objectAtIndex:1];
                    NSNumber* z = [comps objectAtIndex:2];
                    return [NSArray arrayWithObjects: x, y, z, nil];
                }
            }
        }
    }
    NSLog(@"Brown alert: pointing vector not found in camera model.");
    return [NSArray arrayWithObjects:nil];
}
@end

File: MarsImagesIOS/CameraModel.swift
//
//  Model.swift
//  MarsImagesIOS
//
//  Created by Powell, Mark W (397F) on 8/20/17.
//  Copyright © 2017 Mark Powell. All rights reserved.
//

import Foundation
import SwiftyJSON

protocol CameraModel {
    
    var xdim:Double {get set}
    var ydim:Double {get set}
    
    func cmod_2d_to_3d(pos2:[Double], pos3: inout [Double], uvec3: inout [Double])
    
    func cmod_3d_to_2d(pos3:[Double], range: inout [Double], pos2: inout [Double])
}

class CameraModelUtils {
    
    static func model(_ json: JSON) -> CameraModel {
        var returnedModel: CameraModel?
        var type = ""
        var mtype:Double
        var mparm:Double
        var width:Double
        var height:Double
        
        let model = json.arrayValue[1]
        let dimensions = json.arrayValue[0]
        type = model["type"].stringValue
        let comps = model["components"].dictionaryValue
        let cDict = comps["c"]?.arrayValue
        let aDict = comps["a"]?.arrayValue
        let hDict = comps["h"]?.arrayValue
        let vDict = comps["v"]?.arrayValue
        let oDict = comps["o"]?.arrayValue
        let rDict = comps["r"]?.arrayValue
        let eDict = comps["e"]?.arrayValue
        mtype = comps["t"]?.doubleValue ?? 0.0
        mparm = comps["p"]?.doubleValue ?? 0.0
        
        let area = dimensions["area"].arrayValue
        width = area[0].doubleValue
        height = area[1].doubleValue
        
        if type == "CAHV" {
            let cahv = CAHV()
            cahv.c = [cDict![0].doubleValue, cDict![1].doubleValue, cDict![2].doubleValue]
            cahv.a = [aDict![0].doubleValue, aDict![1].doubleValue, aDict![2].doubleValue]
            cahv.h = [hDict![0].doubleValue, hDict![1].doubleValue, hDict![2].doubleValue]
            cahv.v = [vDict![0].doubleValue, vDict![1].doubleValue, vDict![2].doubleValue]
            returnedModel = cahv;
        } else if type == "CAHVOR" {
            let cahvor = CAHVOR()
            cahvor.c = [cDict![0].doubleValue, cDict![1].doubleValue, cDict![2].doubleValue]
            cahvor.a = [aDict![0].doubleValue, aDict![1].doubleValue, aDict![2].doubleValue]
            cahvor.h = [hDict![0].doubleValue, hDict![1].doubleValue, hDict![2].doubleValue]
            cahvor.v = [vDict![0].doubleValue, vDict![1].doubleValue, vDict![2].doubleValue]
            cahvor.o = [oDict![0].doubleValue, oDict![1].doubleValue, oDict![2].doubleValue]
            cahvor.r = [rDict![0].doubleValue, rDict![1].doubleValue, rDict![2].doubleValue]
            returnedModel = cahvor
        } else if type == "CAHVORE" {
            let cahvore = CAHVORE()
            cahvore.c = [cDict![0].doubleValue, cDict![1].doubleValue, cDict![2].doubleValue]
            cahvore.a = [aDict![0].doubleValue, aDict![1].doubleValue, aDict![2].doubleValue]
            cahvore.h = [hDict![0].doubleValue, hDict![1].doubleValue, hDict![2].doubleValue]
            cahvore.v = [vDict![0].doubleValue, vDict![1].doubleValue, vDict![2].doubleValue]
            cahvore.o = [oDict![0].doubleValue, oDict![1].doubleValue, oDict![2].doubleValue]
            cahvore.r = [rDict![0].doubleValue, rDict![1].doubleValue, rDict![2].doubleValue]
            cahvore.e = [eDict![0].doubleValue, eDict![1].doubleValue, eDict![2].doubleValue]
            cahvore.t = mtype
            cahvore.p = mparm
            returnedModel = cahvore
        }
        if var returnedModel = returnedModel {
            returnedModel.xdim = width
            returnedModel.ydim = height
            return returnedModel;
        }
        return CAHV()
    }
    
    static func pointingVector(_ modelJson:JSON) -> [Double] {
        return vectorFromJSON(tripleKey: "camera_vector", modelJson: modelJson)
    }
    
    static func origin(_ modelJson:JSON) -> [Double] {
        return vectorFromJSON(tripleKey: "origin", modelJson: modelJson)
    }
    
    private static func vectorFromJSON(tripleKey:String, modelJson:JSON) -> [Double] {
        let model = modelJson.arrayValue[1]
        let comps = model[tripleKey]
        let x:Double = Double(comps.arrayValue[0].stringValue)!
        let y:Double = Double(comps.arrayValue[1].stringValue)!
        let z:Double = Double(comps.arrayValue[2].stringValue)!
        return [x,y,z]
    }
    
}

